cmd/driver-registrar/main.go

```diff
@@ -110,6 +156,45 @@ func main() {
 	}
 	glog.V(2).Infof("CSI driver node ID: %q", csiDriverNodeId)
 
+	// When kubeletRegistrationPath is specified then driver-registrar ONLY acts
+	// as gRPC server which replies to registration requests initiated by kubelet's
+	// pluginswatcher infrastructure. Node labeling is done by kubelet's csi code.
+	if *kubeletRegistrationPath != "" {
+		registrar := newRegistrationServer()
+		registrar.driverName = csiDriverName
+		registrar.endpoint = *kubeletRegistrationPath
+		registrar.version = supportedVersions
+		socketPath := fmt.Sprintf("/registration/%s-reg.sock", csiDriverName)
+		// Remove any socket, stale or not, but fall through for other files
+		fi, err := os.Stat(socketPath)
+		if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+			os.Remove(socketPath)
+		}
+		if err != nil && !os.IsNotExist(err) {
+			glog.Errorf("failed to stat the socket %s with error: %+v", socketPath, err)
+			os.Exit(1)
+		}
+		// Default to only user accessible socket, caller can open up later if desired
+		oldmask := unix.Umask(0077)
+
+		glog.Infof("Starting Registration Server at: %s\n", socketPath)
+		lis, err := net.Listen("unix", socketPath)
+		if err != nil {
+			glog.Errorf("failed to listen on socket: %s with error: %+v", socketPath, err)
+			os.Exit(1)
+		}
+		unix.Umask(oldmask)
+		glog.Infof("Registration Server started at: %s\n", socketPath)
+		registrar.grpcServer = grpc.NewServer()
+		// Registers kubelet plugin watcher api.
+		registerapi.RegisterRegistrationServer(registrar.grpcServer, registrar)
+
+		// Starts service
+		if err := registrar.grpcServer.Serve(lis); err != nil {
```

Comment:
`Serve returns when lis.Accept fails with fatal errors. lis will be closed when this method returns. Serve will return a non-nil error unless Stop or GracefulStop is called.`

So if the service is gracefully stopped, then `err == nil` and `os.Exit(1)` will not happen, and the code outside the if block will start executing: e.g. `glog.V(1).Infof("Loading kubeconfig.")...)`.

Put a `os.Exit(1)` after this if block to catch that.

Reply:
done